Add Numba void* -> typed-pointer intrinsics and tests

sclaus2 opened the pull request:
This PR adds a small, reliable Numba intrinsic factory for casting void* (FFCx custom_data) to typed C pointers and provides convenience casters for common primitive types, plus unit tests exercising those casters. This PR is relevant for custom_data functionality in dolfinx [https://github.com/FEniCS/dolfinx/pull/4013](url).

sclaus2 committed:
Implement a generic Numba intrinsic factory in utils.py:
create_voidptr_to_dtype_ptr_caster to bitcast void* to typed pointer.
Provide convenience casters for common primitives: float64, float32, int32, int64.
Add unit tests in test_numba_custom_data.py exercising the new casters:
Tests cover basic float/double and int pointer casting and multi-parameter usage.

sclaus2 committed:
ruff and mypy fixes

sclaus2 committed:
align=True instead of padding

schnellerhase commented on ffcx/codegeneration/utils.py:
```diff
@@ -159,3 +159,54 @@ def codegen(context, builder, signature, args):
         sig = numba.types.voidptr(arr)
         return sig, codegen
+
+    def create_voidptr_to_dtype_ptr_caster(target_dtype):
```
Can we add type hints to the function declarations?

schnellerhase commented on ffcx/codegeneration/utils.py:
```diff
@@ -159,3 +159,54 @@ def codegen(context, builder, signature, args):
+        def voidptr_to_dtype_ptr(typingctx, src):
+            # Accept void pointers in various Numba representations:
+            # - CPointer(void): from UFCx cfunc signatures (shows as 'none*')
+            # - voidptr: from numba.cfunc("...(voidptr)") signatures
```
In which case do we need the `voidptr` cast?

sclaus2 committed:
Add Numba intrinsics type hints and improve voidptr caster logic in FFCx utils

ffcx: annotate create_voidptr_to_dtype_ptr_caster with type hints and clarify when void* casts are required.

- Add docstring explaining use cases for voidptr cast (custom_data in UFCx kernels).
- Validate input types early and raise clear TypingError on invalid source types.

sclaus2 commented on ffcx/codegeneration/utils.py:
```diff
@@ -159,3 +159,54 @@ def codegen(context, builder, signature, args):
+                f"got {src}. Ensure you are passing a void* "
+                "(e.g., custom_data from UFCx kernel signature)."
+            )
+            raise numba.core.errors.TypingError(msg)
```
Validated input types early and raise clear TypingError on invalid source types.

jhale commented on ffcx/codegeneration/utils.py:
```diff
@@ -159,3 +159,62 @@ def codegen(context, builder, signature, args):
+
+        return voidptr_to_dtype_ptr
+
+    # Pre-built casters for common types (convenience wrappers)
```
Remove these, trivial and add clutter.

sclaus2 replied:
removed

sclaus2 committed:
Merge remote-tracking branch 'origin/main' into sclaus2/add_voidptr_cast_intrinsic

sclaus2 committed:
Merge remote-tracking branch 'origin/main' into sclaus2/add_voidptr_cast_intrinsic

sclaus2 committed:
address comments

sclaus2 commented on test/test_numba_custom_data.py:
```diff
@@ -0,0 +1,165 @@
+    assert b[0] == pytest.approx(6.5)
+
+
+def test_numba_voidptr_struct_like_mixed_types():
```
ok. done

sclaus2 committed:
update ruff check
